topologies/binding/options.go

```diff
@@ -101,43 +101,48 @@ func (d *dialer) loadCertificates() (*x509.CertPool, tls.Certificate, error) {
 //
 //lint:ignore U1000 will be used by the binding.
 func (d *dialer) dialGRPC(ctx context.Context, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
-	switch {
-	case d.Insecure:
-		tc := insecure.NewCredentials()
-		opts = append(opts, grpc.WithTransportCredentials(tc))
-	case d.SkipVerify:
-		tc := credentials.NewTLS(&tls.Config{InsecureSkipVerify: true})
-		opts = append(opts, grpc.WithTransportCredentials(tc))
-	case d.MutualTls:
-		trusBundle, keyPair, err := d.loadCertificates()
-		if err != nil {
-			return nil, err
+	if len(opts) == 0 {
```

Comment:
Rather than this, can you instead change it so that the provide opts are appended to the default opts? That should cause any transport credentials provided in the opts to overwrite the default credentials.

Reply:
I think something like this might work to allow a test to override whatever the base config sets up:

```go
func (d *dialer) dialGRPC(ctx context.Context, overrideOpts ...grpc.DialOption) (*grpc.ClientConn, error) {
	var opts []grpc.DialOption
	switch {
	case d.Insecure:
		tc := insecure.NewCredentials()
		opts = append(opts, grpc.WithTransportCredentials(tc))
	case d.SkipVerify:
		tc := credentials.NewTLS(&tls.Config{InsecureSkipVerify: true})
		opts = append(opts, grpc.WithTransportCredentials(tc))
	case d.MutualTls:
		trusBundle, keyPair, err := d.loadCertificates()
		if err != nil {
			return nil, err
		}
		tls := &tls.Config{
			Certificates: []tls.Certificate{keyPair},
			RootCAs:      trusBundle,
		}
		tlsConfig := credentials.NewTLS(tls)
		opts = append(opts, grpc.WithTransportCredentials(tlsConfig))
	}
	if d.Username != "" {
		c := &creds{d.Username, d.Password, !d.Insecure}
		opts = append(opts, grpc.WithPerRPCCredentials(c))
	}
	if d.MaxRecvMsgSize != 0 {
		opts = append(opts, grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(int(d.MaxRecvMsgSize))))
	}
	if d.Timeout != 0 {
		retryOpt := grpc_retry.WithPerRetryTimeout(time.Duration(d.Timeout) * time.Second)
		opts = append(opts,
			grpc.WithStreamInterceptor(grpc_retry.StreamClientInterceptor(retryOpt)),
			grpc.WithUnaryInterceptor(grpc_retry.UnaryClientInterceptor(retryOpt)),
		)
		var cancelFunc context.CancelFunc
		ctx, cancelFunc = context.WithTimeout(ctx, time.Duration(d.Timeout)*time.Second)
		defer cancelFunc()
	}
	opts = append(opts, overrideOpts...)
	return grpc.DialContext(ctx, d.Target, opts...)
}
```

Reply:
done

Reply:
Looks like this wasn't quite done right. In bstoll's example overrideOpts is the parameter name, but in what you have, it's declared as a local variable and left empty?

Reply:
my bad, will fix that